core/domain/state_domain.py

```diff
@@ -2852,6 +3085,24 @@ def is_set_data_format(self):
             TranslatableItem.DATA_FORMAT_SET_OF_UNICODE_STRING]
 
 
+DomainObjectCustomizationArgsConversionFnTypes = Union[
+    Callable[[SubtitledHtml, str], SubtitledHtml],
+    Callable[[SubtitledHtml, str], SubtitledHtmlDict],
+    Callable[[SubtitledUnicode, str], SubtitledUnicodeDict],
+    Callable[[SubtitledHtml, str], List[str]]
+]
+
+DictCustomizationArgsConversionFnTypes = Union[
+    Callable[[Dict[str, str], Literal['SubtitledUnicode']], SubtitledUnicode],
+    Callable[[Dict[str, str], Literal['SubtitledHtml']], SubtitledHtml]
+]
+
+AcceptableConversionFnType = Union[
```

Comment:
You should not need this? Each function type should take only one of the above types -- that was the point of the sepration.

Reply:
> Each function type should take only one of the above types

I'm not able to understand what you mean by this because this `AcceptableConversionFnType` is defined for the argument of `traverse_by_schema_and_convert` ( and i think i point it out earlier also ).

https://github.com/oppia/oppia/blob/9980b23608f22a227ea53adff5312b8d4dbfc342/core/domain/state_domain.py#L1351

`AcceptableConversionFnType` is defined for the argument `conversion_fn`

Reply:
Ah, sorry, I should have explained this more. I dug into a bit more detail, here is my thought process.

The traverse function is called in 3 places. Two of them, convert_html_in_interaction() and to_customization_arg_dict() (where the conversion_fn takes a SubtitledHtml/SubtitledUnicode value), seem fine. However, there is also a from_customization_arg_dict() (where the conversion_fn takes a dict value). 

If you read the traverse_by_schema_and_convert() **docstring**, it seems inconsistent with what from_customization_arg_dict() is using it for (the conversion goes in the other direction and the dict wouldn't contain any domain objects).

So there seems to be a contradiction here, in terms of what the various functions are actually supposed to do, that needs to be resolved. This is the source of my confusion. Does this more specific context help clarify?

Reply:
I think this issue now belongs to PR #15870 more than this PR because there we have discussed more about customization_args.